### PR TITLE
Use formatted text view for attributed text editing

### LINKED ISSUE #990: Allow WYSIWYG formatting for annotation text editing
Allow WYSIWYG formatting for annotation text editing (similar to annotation comment editing). Currently the raw text with HTML tags is presented.

### PR DESCRIPTION
**Pull request by mvasilak**
* Adds attributed text editing for item titles
* Adds attributed text editing for annotation text
* Restores placeholder text, if text is empty

Closes #990

### CODE REVIEW

File: Zotero/Scenes/Detail/Annotation Popover/Views/HighlightEditCell.swift [hunk 1]
```diff
@@ -10,49 +10,71 @@ import UIKit
...
-
-    private var observer: AnyObserver<(String, Bool)>?
-    var textObservable: Observable<(String, Bool)> {
-        return Observable.create { observer -> Disposable in
-            self.observer = observer
-            return Disposables.create()
+final class HighlightEditCell: RxTableViewCell {
+    let attributedTextAndHeightReloadNeededObservable: PublishSubject<(NSAttributedString, Bool)>
+
+    private var lineView: UIView?
```
Comment (michalrentka) [inline Zotero/Scenes/Detail/Annotation Popover/Views/HighlightEditCell.swift:16]: Views can still stay `weak`, they are retained by content view.

File: Zotero/Scenes/Detail/Annotation Popover/Views/HighlightEditCell.swift [hunk 2]
```diff
@@ -10,49 +10,71 @@ import UIKit
...
-    func setup(with text: String, color: String) {
+    func setup(with text: NSAttributedString, color: String) {
         let paragraphStyle = NSMutableParagraphStyle()
         paragraphStyle.minimumLineHeight = AnnotationPopoverLayout.annotationLayout.lineHeight
         paragraphStyle.maximumLineHeight = AnnotationPopoverLayout.annotationLayout.lineHeight
-        let attributedText = NSAttributedString(string: text, attributes: [.font: AnnotationPopoverLayout.annotationLayout.font,
-                                                                           .foregroundColor: Asset.Colors.annotationText.color,
-                                                                           .paragraphStyle: paragraphStyle])
+        let attributedText = NSMutableAttributedString(attributedString: text)
+        attributedText.addAttributes([.foregroundColor: Asset.Colors.annotationText.color, .paragraphStyle: paragraphStyle], range: .init(location: 0, length: attributedText.length))
```
Comment (michalrentka) [inline Zotero/Scenes/Detail/Annotation Popover/Views/HighlightEditCell.swift:67]: Same as above, isn't passing the same array in initializer simpler?

File: Zotero/Scenes/Detail/ItemDetail/Views/ItemDetailTitleContentView.swift [hunk 3]
```diff
@@ -7,52 +7,82 @@
...
 
 import UIKit
+import RxSwift
 
 final class ItemDetailTitleContentView: UIView {
-    @IBOutlet private weak var textView: UITextView!
-    @IBOutlet private weak var topConstraint: NSLayoutConstraint!
-    @IBOutlet private weak var bottomConstraint: NSLayoutConstraint!
-    @IBOutlet private weak var separatorHeight: NSLayoutConstraint!
+    private var textView: FormattedTextView!
```
Comment (michalrentka) [inline Zotero/Scenes/Detail/ItemDetail/Views/ItemDetailTitleContentView.swift:13]: `weak` views and constraints (unless they are being deactivated at some point)

File: Zotero/Scenes/Detail/Annotation Popover/Views/AnnotationEditViewController.swift [hunk 4]
```diff
@@ -10,8 +10,7 @@ import UIKit
 
 import RxSwift
 
-// key, color, lineWidth, fontSize, pageLabel, updateSubsequentLabels, highlightText
-typealias AnnotationEditSaveAction = (String, CGFloat, UInt, String, Bool, String) -> Void
+typealias AnnotationEditSaveAction = (_ key: String, _ lineWidth: CGFloat, _ fontSize: UInt, _ pageLabel: String, _ updateSubsequentLabels: Bool, _ highlightText: NSAttributedString) -> Void
```
[the remark this answers is not in the document]
Comment (mvasilak) [reply Zotero/Scenes/Detail/Annotation Popover/Views/AnnotationEditViewController.swift:13]: This is a closure, not a tuple. I just added labels to make it easier to see what is what, when you quick look the definition. Do you suggest we pass one struct as a parameter instead?

File: Zotero/Scenes/Detail/Annotation Popover/Views/AnnotationEditViewController.swift [hunk 5]
```diff
@@ -133,7 +132,8 @@ final class AnnotationEditViewController: UIViewController {
             let paragraphStyle = NSMutableParagraphStyle()
             paragraphStyle.minimumLineHeight = AnnotationPopoverLayout.annotationLayout.lineHeight
             paragraphStyle.maximumLineHeight = AnnotationPopoverLayout.annotationLayout.lineHeight
-            let attributedText = NSAttributedString(string: self.viewModel.state.highlightText, attributes: [.font: AnnotationPopoverLayout.annotationLayout.font, .paragraphStyle: paragraphStyle])
+            let attributedText = NSMutableAttributedString(attributedString: viewModel.state.highlightText)
```
[the remark this answers is not in the document]
Comment (mvasilak) [reply Zotero/Scenes/Detail/Annotation Popover/Views/AnnotationEditViewController.swift:135]: `viewModel.state.highlightText` has changed to an attributed string, converted from HTML, with all the proper fonts. Here we just add the paragraph style.

File: Zotero/Scenes/Detail/Annotation Popover/Views/HighlightEditCell.swift (hunk 2 above)
Comment (mvasilak) [reply Zotero/Scenes/Detail/Annotation Popover/Views/HighlightEditCell.swift:67]: `text` has also changed to an attributed string.

File: Zotero/Scenes/Detail/ItemDetail/Views/ItemDetailTitleContentView.swift (hunk 3 above)
Comment (mvasilak) [reply Zotero/Scenes/Detail/ItemDetail/Views/ItemDetailTitleContentView.swift]: Check https://github.com/zotero/zotero-ios/pull/991/commits/6433d3ff8c672555585e790c4d748ca4f7cbbc6a

File: Zotero/Scenes/Detail/Annotation Popover/Views/HighlightEditCell.swift (hunk 1 above)
Comment (mvasilak) [reply Zotero/Scenes/Detail/Annotation Popover/Views/HighlightEditCell.swift]: Check https://github.com/zotero/zotero-ios/pull/991/commits/6433d3ff8c672555585e790c4d748ca4f7cbbc6a

File: Zotero/Scenes/Detail/Annotation Popover/Views/AnnotationEditViewController.swift (hunk 4 above)
[the remark this answers is not in the document]
Comment (michalrentka) [reply Zotero/Scenes/Detail/Annotation Popover/Views/AnnotationEditViewController.swift:13]: My bad, I didn't notice it's a closure. Anyway we should decide whether we want to convert many parameters or big tuples to structs instead. Maybe >4 => struct? Or is it not worth it in your opinion?

[the remark this answers is not in the document]
Comment (mvasilak) [reply Zotero/Scenes/Detail/Annotation Popover/Views/AnnotationEditViewController.swift]: I'd say it depends on the situation, ideally we should create new structs where these make sense semantically, not just to package passed parameters. Here we can take advantage of `AnnotationEditState.Data`. Check https://github.com/zotero/zotero-ios/pull/991/commits/88b9ccf7265e33a95b6efac9fcac4fd629fe2428

**Review by michalrentka — APPROVED** [review]

**Pull request merged**